### PR TITLE
fix regression that denied access to launch some Apps

### LINKED ISSUE #44553: AWS App Access not working (using integration)
After enabling the AWS App Access for an integration, we are not able to log into it.

The app_server that exists looks like this
```yaml
kind: app_server
metadata:
  name: teleportdev
spec:
  app:
#...
    spec:
      integration: teleportdev
      public_addr: teleportdev.dinis.cloud.gravitational.io:443
      uri: https://console.aws.amazon.com
#...
version: v3
```

When I click to open the AWS Console, I can see that at first the url looks like this
```
https://dinis.cloud.gravitational.io/web/launch/teleportdev.dinis.cloud.gravitational.io:443/dinis.cloud.gravitational.io/teleportdev.dinis.cloud.gravitational.io:443/arn:aws:iam::123456789012:role%2FMarcoTestReadOnlyAccess
```

But then, it loses the port in the URL right after `/launch/` when generating the 2nd URL (the one that has the state)
image

This happens in `master` but I suspect it will be the same in release branches.

Possibly related to https://github.com/gravitational/teleport/pull/43525

Confirmed that removing the check added in the PR above solves the problem.

### PR DESCRIPTION
**Pull request by capnspacehook**
Fixes https://github.com/gravitational/teleport/issues/44553.

changelog: fix regression that denied access to launch some Apps

### CODE REVIEW

**Review by r0mant — APPROVED** [review]
Did you test it with the actual AWS console app?

@kimlisa Can you please review.

**Comment by capnspacehook** [comment]
Yep, tested with AWS console app and multiple variations of normal apps as well

File: web/packages/teleport/src/AppLauncher/AppLauncher.tsx [hunk 1]
```diff
@@ -140,6 +146,19 @@ export function AppLauncherAccessDenied(props: AppLauncherAccessDeniedProps) {
   return <AccessDenied message={props.statusText} />;
 }
 
+// stripFQDN removes the port from the FQDN if it has one and ensures
+// the FQDN is lowercase. This is to prevent issues matching the
+// resolved fqdn with the one that was passed. Apps generally aren't
+// supposed to have a port in the public address but some integrations
+// create apps that do.
+function stripFQDN(fqdn: string) {
+  const portStartIdx = fqdn.indexOf(':');
```
Comment (zmb3) [inline web/packages/teleport/src/AppLauncher/AppLauncher.tsx:155]: Is it possible for an app's FQDN to be an IPv6 address? I don't think this logic would work on an IPv6 address.

Comment (capnspacehook) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx:155]: That's a great question, I'm not sure. I know that apps created from a static config are explicitly not allowed to have IP addresses in the `public_addr` (https://github.com/gravitational/teleport/blob/master/lib/service/servicecfg/app.go#L120) but that check is not enforced for dynamically created apps. I'm pretty sure that's a bug, but I know other app servers (such as the AWS app server) seem to always create apps with a port, which is also disallowed for statically created apps but not dynamically created ones.

I would guess IP addresses aren't allowed in app public addresses, but I'll confirm to be sure.

Comment (webvictim) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx:155]: IP addresses have never historically been allowed in app `public_addrs` AFAIK.

File: web/packages/teleport/src/AppLauncher/AppLauncher.tsx [hunk 2]
```diff
@@ -140,6 +146,19 @@ export function AppLauncherAccessDenied(props: AppLauncherAccessDeniedProps) {
   return <AccessDenied message={props.statusText} />;
 }
 
+// stripFQDN removes the port from the FQDN if it has one and ensures
+// the FQDN is lowercase. This is to prevent issues matching the
+// resolved fqdn with the one that was passed. Apps generally aren't
+// supposed to have a port in the public address but some integrations
+// create apps that do.
+function stripFQDN(fqdn: string) {
```
Comment (kimlisa) [inline web/packages/teleport/src/AppLauncher/AppLauncher.tsx:154]: could we use maybe [new URL](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL)? and we can get the hostname without the port that way.

i don't think the fqdn returns with the protocol (but then again, it's not supposed to include the port either...), so with missing protocol, we can just add one `https://${fqdn}`

[the remark this answers is not in the document]
Comment (capnspacehook) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx]: I renamed it to `prepareFqdn`, as even if I have a `fqdnMatches` function I'll need a sub function to prepare each FQDN to match. Figured I'd just cut out the middleman and use one function

File: web/packages/teleport/src/AppLauncher/AppLauncher.tsx [hunk 3]
```diff
@@ -52,7 +52,13 @@ export function AppLauncher() {
...
         arn: params.arn,
       });
-      if (resolvedApp.fqdn !== params.fqdn) {
+      // Because the ports are stripped from the FQDNs before they are
+      // compared, an attacker can pass a FQDN with a different port than
+      // what the app's public address is configured with and have Teleport
+      // redirect to the public address with an arbitrary port. But because
+      // the attacker can't control what domain is redirected to this has
+      // a low risk factor.
+      if (stripFQDN(resolvedApp.fqdn) !== stripFQDN(params.fqdn)) {
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx]: Renamed to `prepareFqdn` and altered the function docstring to make it more clear what the function is doing.

File: web/packages/teleport/src/AppLauncher/AppLauncher.tsx (hunk 2 above)
Comment (capnspacehook) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx]: For some reason that doesn't work... when you create a URL object from a URL string with a port, the returned URL object's `port` property is an empty string for some reason. So setting it to an empty string to try and remove the port has no effect.

Comment (zmb3) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx]: ??

image

Comment (capnspacehook) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx]: Ok apparently there's different behaviors depending on the JS runtime, what works in Node does not work in Firefox/chromium... I originally set the scheme to `scheme` so if the prepared URLs used to match were accidentally redirected to it wouldn't work. With Node this works fine, but with chromium/Firefox if the scheme isn't recognized the port is set to an empty string

Comment (capnspacehook) [reply web/packages/teleport/src/AppLauncher/AppLauncher.tsx]: The current commit works, I just opted to use `https` as the scheme as @kimlisa originally suggested

File: web/packages/teleport/src/AppLauncher/AppLauncher.test.tsx [hunk 4]
```diff
@@ -97,54 +101,214 @@ describe('app launcher path is properly formed', () => {
...
+    name: 'FQDN, public addr, resolved FQDN with port',
+    path: 'test-app.test.teleport:443/test.teleport/test-app.test.teleport:443?state=ABC',
+    returnedFqdn: 'test-app.test.teleport:443',
+    expectedFqdn: 'test-app.test.teleport:443',
+    expectedPublicAddr: 'test-app.test.teleport',
+    expectedArn: undefined,
+  },
+];
+
+describe('app session request is properly formed', () => {
```
Comment (kimlisa) [inline web/packages/teleport/src/AppLauncher/AppLauncher.test.tsx:259]: i don't think this test is testing what i think you might be trying to test 😅 

you have test cases trying to test with varying uppercases, so you'd probably want the test to fail if the fqdn matching failed, but we aren't testing for the "success" state of render which means any errors will be silent. to test the success state of render:

calling window.replace for redirecting user (we only redirect if there were no errors) and making sure an error wasn't rendered
```js
      await waitFor(() =>
        expect(window.location.replace).toHaveBeenCalled()
      );

     // for any error, we show a "Access Denied" error card
     expect(screen.queryByText(/access denied/i)).not.toBeInTheDocument()
```

we should also add test for error cases. I would specifically test
- non matching fqdns
- invalid url (put some random chars in the fqdn and that should make the `new URL()` throw)

**Review by kimlisa — APPROVED** [review]
@capnspacehook can you review my changes here https://github.com/gravitational/teleport/pull/45024/commits/08e431bb3bd27ceae48454d14779f2227ced3fc1, to make sure i didn't mess up your original logic (i just pulled out common error and wrapped a try catch block)

**Pull request merged**